src/poller.rs

```diff
@@ -50,6 +53,142 @@ impl From<DatabaseError> for PollerError {
     }
 }
 
+fn descriptors(
+    secp: &secp256k1::Secp256k1<impl secp256k1::Verification>,
+    config: &Config,
+    db_vault: &DbVault,
+) -> (
+    DerivedDepositDescriptor,
+    DerivedUnvaultDescriptor,
+    DerivedCpfpDescriptor,
+) {
+    (
+        config
+            .scripts_config
+            .deposit_descriptor
+            .derive(db_vault.derivation_index, secp),
+        config
+            .scripts_config
+            .unvault_descriptor
+            .derive(db_vault.derivation_index, secp),
+        config
+            .scripts_config
+            .cpfp_descriptor
+            .derive(db_vault.derivation_index, secp),
+    )
+}
+
+fn unvault_tx(
+    db_vault: &DbVault,
+    deposit_desc: &DerivedDepositDescriptor,
+    unvault_desc: &DerivedUnvaultDescriptor,
+    cpfp_desc: &DerivedCpfpDescriptor,
+) -> Result<UnvaultTransaction, revault_tx::error::TransactionCreationError> {
+    let deposit_txo = DepositTxOut::new(db_vault.amount, deposit_desc);
+    let deposit_txin = DepositTxIn::new(db_vault.deposit_outpoint, deposit_txo);
+    UnvaultTransaction::new(
+        deposit_txin,
+        &unvault_desc,
+        &cpfp_desc,
+        /* FIXME: remove from the API */ 0,
+    )
+}
+
+fn manage_cancel_attempts(
+    db_path: &path::Path,
+    secp: &secp256k1::Secp256k1<impl secp256k1::Verification>,
+    config: &Config,
+    bitcoind: &BitcoinD,
+    current_tip: &ChainTip,
+) -> Result<(), PollerError> {
+    let canceling_vaults = db_canceling_vaults(db_path)?;
+
+    for db_vault in canceling_vaults {
+        let (deposit_desc, unvault_desc, cpfp_desc) = descriptors(secp, config, &db_vault);
+        let unvault_tx = match unvault_tx(&db_vault, &deposit_desc, &unvault_desc, &cpfp_desc) {
+            Ok(tx) => tx,
+            Err(e) => {
+                // TODO: handle dust better (they should never send us dust vaults though)
+                log::error!("Unexpected error deriving Unvault transaction: '{}'", e);
+                continue;
+            }
+        };
+        let unvault_txin = unvault_tx.revault_unvault_txin(&unvault_desc);
+        let unvault_outpoint = unvault_txin.outpoint();
+        let cancel_tx = CancelTransaction::new(
+            unvault_txin,
+            None,
+            &deposit_desc,
+            /* FIXME: remove from the API */ 0,
+        );
+
+        // If it was confirmed, check for how long and maybe forget it. Otherwise check if it got
+        // confirmed since our last poll.
+        if let Some(conf_height) = db_vault.revoc_height {
+            let n_confs = current_tip
+                .height
+                .checked_add(1)
```

Comment:
why are you adding 1 here? is it because `current_tip` hasn't been updated yet to the _actual_ chain tip?

Reply:
No. It's because you start counting at `0` but you count confirmations from `1`:
```python
if confirmed_block == current_block:
    return 1
if confirmed_block == current_block - 1:
    return 2
if confirmed_block == current_block - 2:
    return 3
```
Therefore it's the number of confirmation is equal to `current_block - confirmed_block + 1`.

Reply:
ah yeah of course :)